docs/fomega/cek-cps-experiments/src/CekMachine.recursive.hs

```diff
@@ -0,0 +1,138 @@
+-- a lot of !s inserted to (hopefully) make sure that
+-- we're evaluating the code eagerly.  This can also
+-- be regarded as the CPS version with continuations
+-- converted back into ordinary function calls.
+
+
+-- | The CEK machine.
+-- Rules are the same as for the CK machine from "Language.PlutusCore.Evaluation.CkMachine",
+-- except we do not use substitution and use environments instead.
+-- The CEK machine relies on variables having non-equal 'Unique's whenever they have non-equal
+-- string names. I.e. 'Unique's are used instead of string names, so the renamer pass is required.
+-- This is for efficiency reasons.
+-- The type checker pass is required as well (and in our case it subsumes the renamer pass).
+-- Feeding ill-typed terms to the CEK machine will likely result in a 'MachineException'.
+-- The CEK machine generates booleans along the way which might contain globally non-unique 'Unique's.
+-- This is not a problem as the CEK machines handles name capture by design.
+
+module Language.PlutusCore.Interpreter.CekMachine
+    ( EvaluationResult (..)
+    , evaluateCek
+    , runCek
+    ) where
+
+import           Language.PlutusCore
+import           Language.PlutusCore.Constant
+import           Language.PlutusCore.Evaluation.MachineException (MachineError (..),
+                                                                  MachineException (..))
+import           Language.PlutusCore.Evaluation.Result           (EvaluationResult (..))
+import           Language.PlutusCore.View
+import           PlutusPrelude
+
+import qualified Data.Text                                       as T
+import qualified Language.PlutusCore.Pretty                      as PLC
+
+import           Data.IntMap                                     (IntMap)
+import qualified Data.IntMap                                     as IntMap
+
+termStr :: Plain Term -> String
+termStr = T.unpack . PLC.prettyPlcDefText
+
+type Plain f = f TyName Name ()
+
+-- | A 'Value' packed together with the environment it's defined in.
+data Closure = Closure
+    { _closureValue       :: Plain Value
+    , _closureEnvironment :: Environment
+    }
+
+-- | Environments used by the CEK machine.
+-- Each row is a mapping from the 'Unique' representing a variable to a 'Closure'.
+newtype Environment = Environment (IntMap Closure)
+
+-- | Extend an environment with a variable name, the value the variable stands for
+-- and the environment the value is defined in.
+extendEnvironment :: Name () -> Plain Value -> Environment -> Environment -> Environment
+extendEnvironment argName arg argEnv (Environment oldEnv) =
+    Environment $ IntMap.insert (unUnique $ nameUnique argName) (Closure arg argEnv) oldEnv
+
+-- | Look up a name in an environment.
+lookupName :: Name () -> Environment -> Maybe Closure
+lookupName name (Environment env) = IntMap.lookup (unUnique $ nameUnique name) env
+
+
+evalCek :: Closure -> Closure
+evalCek !cl@(Closure term env) =
```

Comment:
You force `cl` here by matching on it, hence `!` does not add anything.

Reply:
1. I forgot to mention this, but if you look in RUNNING.md it tells you how to run the tests.

2. I tend to work with R by plotting things interactively and then hacking together a script to do it properly with labels on the axes and a legend and so on once I know what the data looks like.  In this case, it happens that the results for the different experiments are similar enough that you can use more or less the same parameters for each graph, but in general that wouldn't happen.  I think that unifying the scripts is maybe more effort than it's worth, since these particular things tend not to be too resuable (also, the R language is remarkably horrible).  It may be overkill using R here, since it's really for doing statistical stuff like fitting models (and almost anything else you can think of), but I happen to be familiar with it so it was the easiest thing for me to use.  

3. Yes, I think that's about as far as I want to go with writing my own Plutus Core; PlutusTx or PIR would probably be a lot better for generating complex benchmarks (and I think that's something we should do some time).

5. I just stuck in bangs everywhere to be absolutely sure that it was doing things in the right order, but yes, many could be omitted.  I tried a version without any bangs, but I don't think it  made any noticeable difference.

6. The CPS version was quite difficult to get right since you've just got a big chain of continuations and if something goes wrong it's very hard to work out what's wrong. It's much easier to see what's going on with the stack of frames, and maybe a positive outcome of this experiment is that it tells us that using that version isn't actually costing us a lot.

7. I also don't know why the current version is using less memory.  It would be interesting to find out ...